src/Stratis.Bitcoin.Features.RPC.Tests/Controller/FullNodeControllerTest.cs

```diff
@@ -0,0 +1,769 @@
+                this.controller.ValidateAddress("invalidaddress");
+            });
+        }
+
+        [Fact]
+        public void ValidateAddress_ValidAddressOfDifferentNetwork_ReturnsFalse()
+        {
+            // P2PKH
+            var address = new Key().PubKey.GetAddress(Network.Main);
+
+            var result = this.controller.ValidateAddress(address.ToString());
+
+            var isValid = result["isvalid"].ToObject<bool>();
+            Assert.False(isValid);
+        }
+
+        [Fact]
+        public void ValidateAddress_ValidP2PKHAddress_ReturnsTrue()
+        {
+            // P2PKH
+            var address = new Key().PubKey.GetAddress(this.network);
+
+            var result = this.controller.ValidateAddress(address.ToString());
+
+            var isValid = result["isvalid"].ToObject<bool>();
+            Assert.True(isValid);
+        }
+
+        [Fact]
+        public void ValidateAddress_ValidP2SHAddress_ReturnsTrue()
+        {
+            // P2SH
+            var address = new Key().ScriptPubKey.GetScriptAddress(this.network);
+
+            var result = this.controller.ValidateAddress(address.ToString());
+
+            var isValid = result["isvalid"].ToObject<bool>();
+            Assert.True(isValid);
+        }
+
+        [Fact]
+        public void ValidateAddress_ValidP2WPKHAddress_ReturnsTrue()
+        {
+            // P2WPKH
+            var address = new Key().PubKey.WitHash.GetAddress(this.network);
+
+            var result = this.controller.ValidateAddress(address.ToString());
+
+            var isValid = result["isvalid"].ToObject<bool>();
+            Assert.True(isValid);
+        }
+
+        [Fact]
+        public void ValidateAddress_ValidP2WSHAddress_ReturnsTrue()
+        {
+            // P2WSH
+            var address = new Key().PubKey.ScriptPubKey.WitHash.ScriptPubKey.GetWitScriptAddress(this.network);
+
+            var result = this.controller.ValidateAddress(address.ToString());
+
+            var isValid = result["isvalid"].ToObject<bool>();
+            Assert.True(isValid);
+        }
+
+        private Transaction CreateTransaction()
+        {
+            var transaction = new Transaction();
+            transaction.AddInput(TxIn.CreateCoinbase(23523523));
+            transaction.AddOutput(new TxOut(this.network.GetReward(23523523), new Key().ScriptPubKey));
+            return transaction;
+        }
+
+        private string GetBlockHeaderBits(BlockHeader header)
+        {
+
+            byte[] bytes = this.GetBytes(header.Bits.ToCompact());
+            return Encoders.Hex.EncodeData(bytes);
+        }
+
+        private byte[] GetBytes(uint compact)
+        {
+            return new byte[]
+            {
+                (byte)(compact >> 24),
+                (byte)(compact >> 16),
+                (byte)(compact >> 8),
+                (byte)(compact)
+            };
+        }
+
+        private class TestReadOnlyNetworkPeerCollection : IReadOnlyNetworkPeerCollection
```

Comment:
This class is a copy-paste from the 'WalletControllerTest'. Maybe add reference to `Stratis.Bitcoin.Features.Wallet.Tests` and reuse existing `TestReadOnlyNetworkPeerCollection`?

Reply:
Deleted this instance on next commit.